src/Plutarch/Test/Precompiled.hs

```diff
@@ -0,0 +1,179 @@
+import Data.Foldable (toList)
+import Data.Tagged (Tagged (Tagged))
+import Data.Text (Text)
+import Plutarch.Evaluate (EvalError, evalScript)
+import Plutarch.Extra.DebuggableScript (DebuggableScript (DebuggableScript, debugScript, script), mustCompileD)
+import Plutarch.Prelude (ClosedTerm, S, Type)
+import PlutusLedgerApi.V1.Scripts (Script, applyArguments)
+import PlutusLedgerApi.V2 (Data)
+import Test.Tasty (testGroup)
+import Test.Tasty.Providers (IsTest, Result, TestTree, run, singleTest, testFailed, testOptions, testPassed)
+import Text.PrettyPrint (
+    Doc,
+    Style,
+    hang,
+    lineLength,
+    renderStyle,
+    style,
+    vcat,
+ )
+import Text.Show.Pretty (ppDoc)
+
+applyDebuggableScript :: DebuggableScript -> [Data] -> DebuggableScript
+applyDebuggableScript (DebuggableScript script debugScript) d =
+    DebuggableScript
+        { script = applyArguments script d
+        , debugScript = applyArguments debugScript d
+        }
+
+-- | @since 0.1.1.0
+data Expectation
+    = Success
+    | Failure
+    deriving stock (Show)
+
+{- | Holds necessary information for each test cases.
+
+ @since 0.1.1.0
+-}
+data TestCase = TestCase
+    { dScript :: DebuggableScript
+    , caseName :: String
+    , expectation :: Expectation
+    }
+
+ourStyle :: Style
+ourStyle = style{lineLength = 80}
+
+ppLogs :: [Text] -> Doc
+ppLogs = \case
+    [] -> "No logs found. Did you forget to build with +development?"
+    logs -> vcat $ ppDoc <$> logs
+
+instance IsTest TestCase where
+    testOptions = Tagged []
+    run _ TestCase{..} _ = return $
+        case (r, expectation) of
+            (Right _, Success) -> testPassed ""
+            (Right x, Failure) -> failWithStyle . unexpectedSuccess x $ dt
+            (Left err, Success) -> failWithStyle . unexpectedFailure err $ dt
+            (Left _, Failure) -> testPassed ""
+      where
+        (r, _, _) = evalScript $ script dScript
+        (_, _, dt) = evalScript $ debugScript dScript
+        failWithStyle :: Doc -> Result
+        failWithStyle = testFailed . renderStyle ourStyle
+        unexpectedFailure :: EvalError -> [Text] -> Doc
+        unexpectedFailure err logs =
+            "Expected a successful run, but failed instead.\n"
+                <> hang "Error" 4 (ppDoc err)
+                <> hang "Logs" 4 (ppLogs logs)
+        unexpectedSuccess :: Script -> [Text] -> Doc
+        unexpectedSuccess result logs =
+            "Expected a failing run, but succeeded instead.\n"
+                <> hang "Result" 4 (ppDoc result)
+                <> hang "Logs" 4 (ppLogs logs)
+
+{- | Allows monadically defining a test tree that uses same precompiled script.
+
+ @since 0.1.1.0
+-}
+newtype TestCompiled a = TestCompiled
+    { unTestCompiled ::
+        RWS DebuggableScript (Acc TestCase) () a
+    }
+    deriving
+        ( Functor
+        , Applicative
+        , Monad
+        , MonadReader DebuggableScript
+        , MonadWriter (Acc TestCase)
```

Comment:
`since`s on all these derivations. I also don't know if having a `MonadWriter` derivation on this is a good idea: it has a lot of functionality, and I would argue most people would only need to do 'add one test'. That would be better as a dedicated function.

Reply:
I don't see it particularly harmful. Besides, I might later use this to add ability to add batch of tests.

Reply:
Still requires `since`s.